File: fsharp/ledger/Ledger.fs
module Ledger

open System
open System.Globalization

// just left the Entry type and its 'constructor' as is, since test cases depend on them
type public Entry = { dat: DateTime; des: string; chg: int } // agnostic to locale and currency 

let public mkEntry (date: string) description change = 
    { dat = DateTime.Parse(date, CultureInfo.InvariantCulture); des = description; chg = change }

// add types to get complete pattern matches 
type private Locale   = US     | NL
type private Currency = Dollar | Euro 

// formatting constants
let private dateWidth = 10
let private descrWidth = 25
let private changeWidth = 13
let private separator = " | "

// formatting functions
let private header = function
    | US -> "Date".PadRight(dateWidth)  + separator + "Description".PadRight(descrWidth)  + separator + "Change".PadRight(changeWidth)
    | NL -> "Datum".PadRight(dateWidth) + separator + "Omschrijving".PadRight(descrWidth) + separator + "Verandering".PadRight(changeWidth)

let private fmtDate (dat: DateTime) = function 
    | US -> (dat.ToString("MM\/dd\/yyyy")).PadRight(dateWidth)
    | NL -> (dat.ToString("dd-MM-yyyy")).PadRight(dateWidth)

let private fmtDescr (des: string) = 
    match des with
    | _ when des.Length <= descrWidth -> des.PadRight(descrWidth)
    | _                               -> des.Substring(0, descrWidth - 3) + "..."

let private fmtAmount (change: int) locale currency =
    // collect some variables
    let cultInfo = if locale = US then (new CultureInfo("en-US")) else (new CultureInfo("nl-NL"))
    let symb = 
        match currency with
        | Dollar -> "$"
        | Euro   -> "€"
    let amt = (float change) / 100.00
    let isNegative = (amt < 0)
    // get different parts of the string
    let front =  // negative amounts are wrapped in braces in US locale, NL locale likes a space
        match (isNegative, locale) with 
        | (false, US) ->       symb
        | (true,  US) -> "(" + symb
        | (_,     NL) ->       symb + " "
    let formattedDigits = 
        match isNegative with // get rid of negative sign in US locale
        | true when locale = US -> amt.ToString("#,#0.00", cultInfo).Substring(1)
        | _                     -> amt.ToString("#,#0.00", cultInfo)
    let back = // there is no real logic behind the first case, I think
        match (isNegative, locale) with 
        | (false, _ ) -> " "
        | (true,  NL) -> ""
        | (true,  US) -> ")"
    // concat and apply padding
    (front + formattedDigits + back).PadLeft(changeWidth)

// use all this to format a complete entry dependent on locale and currency
let private fmtEntry (entry: Entry) locale currency =
    fmtDate entry.dat locale + separator 
    + fmtDescr entry.des     + separator 
    + fmtAmount entry.chg locale currency

// final public method
let public formatLedger currencyString localeString entries =
    // find the Locale and Currency types from the strings
    let locale   = if localeString   = "en-US" then US     else NL 
    let currency = if currencyString = "USD"   then Dollar else Euro 
    // get a list of strings for the sorted entries
    let lineList = 
        entries 
        |> List.sortBy (fun x -> x.dat, x.des, x.chg)
        |> List.map (fun entry -> fmtEntry entry locale currency) 
    // prepend the header and concat with newlines
    (header locale) :: lineList |> String.concat System.Environment.NewLine

File: csharp/ledger/Ledger.cs
#pragma warning disable CS8524 // suppresses warnings because of non-exhaustive switch statements

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;

// unchanged
public class LedgerEntry
{
   public LedgerEntry(DateTime date, string desc, decimal chg)
   {
       Date = date;
       Desc = desc;
       Chg = chg;
   }

   public DateTime Date { get; }
   public string Desc { get; }
   public decimal Chg { get; }
}


public static class Ledger
{
    // unchanged
    public static LedgerEntry CreateEntry(string date, string desc, int chng) =>
        new LedgerEntry(DateTime.Parse(date, CultureInfo.InvariantCulture), desc, chng / 100.0m);
    

    // It's really a pity that C# enums won't exhaust
    // switch statements (as do the discriminated unions in F#),
    // since the underlying type here is an int and one could also define 
    // something like "LocaleType breakingThings = (LocaleType)5;"
    // I suppress the compiler warnings, since all this is private and 
    // the strings for locale and currency given from the outside are both checked!
    // Theres an interesting discussion going on around these issues, see
    // https://github.com/dotnet/csharplang/discussions/2671
    private enum LocaleType
    {
        US,
        NL
    }
    // saves a CultureInfo object that keeps all necessary information for formatting
    private class Locale
    {
        public Locale(string localeString, string currencyString)
        {
            // check incoming localeString and instantiate CultureInfo object
            Culture = localeString switch
            {
                "en-US" or "nl-NL" => new CultureInfo(localeString),
                 _ => throw new ArgumentException("Invalid locale")
            };
            // check incoming currencyString and set CurrencySymbol on object
            Culture.NumberFormat.CurrencySymbol = currencyString switch
            {
                "USD" => "$",
                "EUR" => "€",
                _ => throw new ArgumentException("Invalid currency")
            };
            // LocType can be found by a ternary, will be used to set further formatting variables
            LocType = (localeString == "en-US" ? LocaleType.US : LocaleType.NL);
            // compiler warnings suppressed for these switch statements
            Culture.NumberFormat.CurrencyNegativePattern = LocType switch
            {
                LocaleType.US => 0,
                LocaleType.NL => 12
            };
            Culture.DateTimeFormat.ShortDatePattern = LocType switch
            {
                LocaleType.US => "MM/dd/yyyy",
                LocaleType.NL => "dd/MM/yyyy"
            };
        }
        public LocaleType LocType { get; }
        public CultureInfo Culture { get; }
    }
    
    // locale-dependent header
    private static string Header(LocaleType loc) => loc switch // compiler warning suppressed
    {
        LocaleType.US => "Date       | Description               | Change       ",
        LocaleType.NL => "Datum      | Omschrijving              | Verandering  "
    };

    // formatting constants and functions almost unchanged
    private const string Separator = " | ";
    private static string FormatDate(IFormatProvider Culture, DateTime date) => date.ToString("d", Culture);
    private static string FormatDescription(string desc) => (desc.Length > 25) ? desc.Substring(0, 22) + "..." : desc.PadRight(25);
    private static string FormatChange(IFormatProvider Culture, decimal cgh) =>
        (cgh < 0.0m ? cgh.ToString("C", Culture) : cgh.ToString("C", Culture) + " ").PadLeft(13);
    private static string FormatEntry(IFormatProvider Culture, LedgerEntry entry) =>
        $"{FormatDate(Culture, entry.Date)}{Separator}{FormatDescription(entry.Desc)}{Separator}{FormatChange(Culture, entry.Chg)}";
    
    // sorting function looks really better this way!
    private static IEnumerable<LedgerEntry> sort(LedgerEntry[] entries) =>
        entries.OrderBy(e => e.Date).ThenBy(e => e.Desc).ThenBy(e => e.Chg);

    // decisive public method with LinQ
    public static string Format(string currencyString, string localeString, LedgerEntry[] entries)
    {
        Locale locale = new Locale(localeString, currencyString);
        return string.Join("\n", 
            sort(entries)
            .Select(entry => FormatEntry(locale.Culture, entry))
            .Prepend(Header(locale.LocType))
        );
    }
}
